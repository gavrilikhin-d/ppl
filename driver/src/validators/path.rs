use std::path::PathBuf;

use miette::{Diagnostic, SourceSpan};
use thiserror::Error;

/// Diagnostic for a wrong path
#[derive(Error, Debug, Diagnostic)]
#[error("path doesn't exist")]
#[diagnostic(code(driver::wrong_path))]
pub struct WrongPath {
    /// Span to the path
    #[label("path doesn't exist")]
    at: SourceSpan,
}

/// Error's that may appear in [`existing_path`] function
#[derive(Error, Diagnostic, Debug)]
pub enum ExistingPathError {
    #[error(transparent)]
    #[diagnostic(code(driver::io_error))]
    IoError(#[from] std::io::Error),

    #[error(transparent)]
    WrongPath(#[from] WrongPath),
}

/// Check that path exists
pub fn exists(path: &str) -> Result<PathBuf, ExistingPathError> {
    let path = PathBuf::from(path);
    if path.try_exists()? {
        return Ok(path);
    }
    Err(WrongPath { at: (0..0).into() }.into())
}

/// Diagnostic for a path, that's not a directory
#[derive(Error, Debug, Diagnostic)]
#[error("expected path to directory")]
#[diagnostic(code(driver::expected_directory))]
pub struct ExpectedDirectory {
    /// Span to the path
    #[label("isn't a directory")]
    at: SourceSpan,
}

/// Error's that may appear in [`existing_dir`] function
#[derive(Error, Diagnostic, Debug)]
pub enum ExistingDirError {
    #[error(transparent)]
    ExistingPathError(#[from] ExistingPathError),

    #[error(transparent)]
    ExpectedDirectory(#[from] ExpectedDirectory),
}

/// Check that path exists and is a directory
pub fn is_dir(path: &str) -> Result<PathBuf, ExistingDirError> {
    let p = exists(path)?;
    if p.is_dir() {
        return Ok(p);
    }
    Err(ExpectedDirectory { at: (0, 0).into() }.into())
}
