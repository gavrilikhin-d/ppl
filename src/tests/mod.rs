use crate::e2es;

e2es! {
    array,
    candidate_not_viable,
    clone,
    common_functions,
    constraints,
    constraints_in_constructor,
    consume_greater,
    deps,
    deref_member_ref,
    destructor,
    empty_block,
    empty_constructor,
    escaped_id,
    generics,
    import_all,
    integer,
    integer_not_eq_rational,
    invalid_indentation,
    memory,
    missing_fields,
    monomorphize,
    monomorphize_predeclared,
    multifile,
    multiple_errors,
    multiple_initialization,
    non_class_constructor,
    plus_assign,
    predeclare_function,
    predeclare_vars,
    rational,
    reference_mut,
    reference_to_none,
    references,
    specify_variable_ty,
    star,
    string,
    supertraits,
    traits,
    type_as_value,
    trait_with_ref,
    type_of,
    wrong_initializer_type
}

#[test]
fn ppl() {
    use std::path::Path;

    use insta::assert_snapshot;
    use tempdir::TempDir;

    // Compile-time check that file exists
    include_bytes!(concat!(env!("CARGO_MANIFEST_DIR"), "/ppl/src/lib.ppl"));

    let temp_dir = TempDir::new("ppl").unwrap();
    let tmp = temp_dir.path();
    let name = "ppl";
    let dir = Path::new(concat!(env!("CARGO_MANIFEST_DIR"), "/ppl"));

    let res = crate::e2e::internal::compile(&tmp, dir);
    if let Err(err) = res {
        assert_snapshot!("ppl.error", err);
        return;
    }

    let hir = crate::e2e::internal::hir(&tmp, name, &dir);
    assert_snapshot!("ppl.hir", hir);

    let ir = crate::e2e::internal::ir(&tmp, name, &dir);
    assert_snapshot!("ppl.ir", ir);
}
