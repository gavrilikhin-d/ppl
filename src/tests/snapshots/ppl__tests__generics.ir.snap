---
source: src/tests/mod.rs
expression: ir
---
; ModuleID = 'main'
source_filename = "src/main.ppl"

%"Type<String>" = type { %String, %Integer }
%String = type { ptr }
%Integer = type { ptr }
%"Point<Integer>" = type { %Integer, %Integer }

@"Type<String>" = private global %"Type<String>" zeroinitializer
@0 = private unnamed_addr constant [7 x i8] c"String\00", align 1
@x = global %Integer zeroinitializer
@1 = private unnamed_addr constant [6 x i8] c"hello\00", align 1

define private void @initialize() !dbg !3 {
  %1 = alloca %"Type<String>", align 8, !dbg !7
  %"Type<String>.name" = getelementptr inbounds %"Type<String>", ptr %1, i32 0, i32 0, !dbg !7
  %2 = call %String @string_from_c_string_and_length(ptr @0, i64 6), !dbg !8
  store %String %2, ptr %"Type<String>.name", align 8, !dbg !8
  %"Type<String>.size" = getelementptr inbounds %"Type<String>", ptr %1, i32 0, i32 1, !dbg !8
  %3 = call %Integer @integer_from_i64(i64 8), !dbg !8
  store %Integer %3, ptr %"Type<String>.size", align 8, !dbg !8
  %4 = load %"Type<String>", ptr %1, align 8, !dbg !8
  store %"Type<String>" %4, ptr @"Type<String>", align 8, !dbg !8
  br label %return, !dbg !8

return:                                           ; preds = %0
  ret void
}

declare %String @string_from_c_string_and_length(ptr, i64)

declare %Integer @integer_from_i64(i64)

define private void @initialize.1() !dbg !9 {
  %1 = alloca %"Point<Integer>", align 8, !dbg !10
  %"Point<Integer>.x" = getelementptr inbounds %"Point<Integer>", ptr %1, i32 0, i32 0, !dbg !10
  %2 = call %Integer @integer_from_i64(i64 0), !dbg !11
  store %Integer %2, ptr %"Point<Integer>.x", align 8, !dbg !11
  %"Point<Integer>.y" = getelementptr inbounds %"Point<Integer>", ptr %1, i32 0, i32 1, !dbg !11
  %3 = call %Integer @integer_from_i64(i64 0), !dbg !12
  store %Integer %3, ptr %"Point<Integer>.y", align 8, !dbg !12
  %4 = load %"Point<Integer>", ptr %1, align 8, !dbg !12
  %5 = call %Integer @"x of <:Point<Integer>>"(%"Point<Integer>" %4), !dbg !12
  store %Integer %5, ptr @x, align 8, !dbg !12
  br label %return, !dbg !12

return:                                           ; preds = %0
  ret void
}

define private %Integer @"x of <:Point<Integer>>"(%"Point<Integer>" %0) !dbg !13 {
  %return_value = alloca %Integer, align 8
  %p = alloca %"Point<Integer>", align 8
  store %"Point<Integer>" %0, ptr %p, align 8
  %x = getelementptr inbounds %"Point<Integer>", ptr %p, i32 0, i32 0, !dbg !14
  %2 = load %Integer, ptr %x, align 8, !dbg !14
  store %Integer %2, ptr %return_value, align 8, !dbg !14
  br label %return, !dbg !14

return:                                           ; preds = %1
  %3 = load %Integer, ptr %return_value, align 8
  ret %Integer %3
}

define void @main.execute() !dbg !15 {
init_globals:
  call void @initialize(), !dbg !16
  call void @initialize.1(), !dbg !17
  br label %0, !dbg !17

0:                                                ; preds = %init_globals
  %1 = load %Integer, ptr @x, align 8, !dbg !18
  %2 = call %Integer @clone_integer(%Integer %1), !dbg !18
  %3 = call %Integer @integer_from_i64(i64 0), !dbg !19
  %4 = call i1 @integer_eq_integer(%Integer %2, %Integer %3), !dbg !19
  call void @"println <:Bool>"(i1 %4), !dbg !19
  %5 = call %Integer @integer_from_i64(i64 1), !dbg !20
  %6 = call %Integer @"id <:Integer>"(%Integer %5), !dbg !20
  %7 = call %Integer @integer_from_i64(i64 1), !dbg !21
  %8 = call i1 @integer_eq_integer(%Integer %6, %Integer %7), !dbg !21
  call void @"println <:Bool>"(i1 %8), !dbg !21
  %9 = call %String @string_from_c_string_and_length(ptr @1, i64 5), !dbg !22
  %10 = call %String @"id <:String>"(%String %9), !dbg !22
  call void @"println <:String>"(%String %10), !dbg !22
  %11 = alloca %"Point<Integer>", align 8, !dbg !23
  %"Point<Integer>.x" = getelementptr inbounds %"Point<Integer>", ptr %11, i32 0, i32 0, !dbg !23
  %12 = call %Integer @integer_from_i64(i64 0), !dbg !24
  store %Integer %12, ptr %"Point<Integer>.x", align 8, !dbg !24
  %"Point<Integer>.y" = getelementptr inbounds %"Point<Integer>", ptr %11, i32 0, i32 1, !dbg !24
  %13 = call %Integer @integer_from_i64(i64 0), !dbg !25
  store %Integer %13, ptr %"Point<Integer>.y", align 8, !dbg !25
  %14 = load %"Point<Integer>", ptr %11, align 8, !dbg !25
  %15 = call %"Point<Integer>" @"id <:Point<Integer>>"(%"Point<Integer>" %14), !dbg !25
  br label %return, !dbg !17

return:                                           ; preds = %0
  ret void
}

define private void @"println <:Bool>"(i1 %0) !dbg !26 {
  %x = alloca i1, align 1
  store i1 %0, ptr %x, align 1
  %2 = load i1, ptr %x, align 1, !dbg !27
  %3 = call %String @"String from <:Bool>"(i1 %2), !dbg !27
  call void @"println <:String>"(%String %3), !dbg !27
  br label %return, !dbg !27

return:                                           ; preds = %1
  ret void
}

declare void @"println <:String>"(%String)

declare %String @"String from <:Bool>"(i1)

declare i1 @integer_eq_integer(%Integer, %Integer)

declare %Integer @clone_integer(%Integer)

define private %Integer @"id <:Integer>"(%Integer %0) !dbg !28 {
  %return_value = alloca %Integer, align 8
  %x = alloca %Integer, align 8
  store %Integer %0, ptr %x, align 8
  %2 = load %Integer, ptr %x, align 8, !dbg !29
  store %Integer %2, ptr %return_value, align 8, !dbg !29
  br label %return, !dbg !29

return:                                           ; preds = %1
  %3 = load %Integer, ptr %return_value, align 8
  ret %Integer %3
}

define private %String @"id <:String>"(%String %0) !dbg !30 {
  %return_value = alloca %String, align 8
  %x = alloca %String, align 8
  store %String %0, ptr %x, align 8
  %2 = load %String, ptr %x, align 8, !dbg !31
  store %String %2, ptr %return_value, align 8, !dbg !31
  br label %return, !dbg !31

return:                                           ; preds = %1
  %3 = load %String, ptr %return_value, align 8
  ret %String %3
}

define private %"Point<Integer>" @"id <:Point<Integer>>"(%"Point<Integer>" %0) !dbg !32 {
  %return_value = alloca %"Point<Integer>", align 8
  %x = alloca %"Point<Integer>", align 8
  store %"Point<Integer>" %0, ptr %x, align 8
  %2 = load %"Point<Integer>", ptr %x, align 8, !dbg !33
  store %"Point<Integer>" %2, ptr %return_value, align 8, !dbg !33
  br label %return, !dbg !33

return:                                           ; preds = %1
  %3 = load %"Point<Integer>", ptr %return_value, align 8
  ret %"Point<Integer>" %3
}

!llvm.module.flags = !{!0}
!llvm.dbg.cu = !{!1}

!0 = !{i32 2, !"Debug Info Version", i32 3}
!1 = distinct !DICompileUnit(language: DW_LANG_C, file: !2, producer: "ppl", isOptimized: false, runtimeVersion: 0, emissionKind: FullDebug, splitDebugInlining: false, sysroot: "/")
!2 = !DIFile(filename: "src/main.ppl", directory: ".")
!3 = distinct !DISubprogram(name: "initialize", linkageName: "initialize", scope: !2, file: !2, line: 13, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!4 = !DISubroutineType(types: !5)
!5 = !{!6}
!6 = !DIBasicType(name: "i32", size: 32, encoding: DW_ATE_signed)
!7 = !DILocation(line: 13, scope: !3)
!8 = !DILocation(line: 0, scope: !3)
!9 = distinct !DISubprogram(name: "initialize.1", linkageName: "initialize.1", scope: !2, file: !2, line: 5, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!10 = !DILocation(line: 5, column: 13, scope: !9)
!11 = !DILocation(line: 5, column: 24, scope: !9)
!12 = !DILocation(line: 5, column: 30, scope: !9)
!13 = distinct !DISubprogram(name: "x of <:Point<Integer>>", linkageName: "x of <:Point<Integer>>", scope: !9, file: !2, line: 3, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!14 = !DILocation(line: 3, column: 28, scope: !13)
!15 = distinct !DISubprogram(name: "main.execute", linkageName: "main.execute", scope: !2, file: !2, line: 6, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!16 = !DILocation(line: 13, scope: !15)
!17 = !DILocation(line: 5, column: 8, scope: !15)
!18 = !DILocation(line: 6, column: 8, scope: !15)
!19 = !DILocation(line: 6, column: 13, scope: !15)
!20 = !DILocation(line: 10, column: 12, scope: !15)
!21 = !DILocation(line: 10, column: 18, scope: !15)
!22 = !DILocation(line: 11, column: 12, scope: !15)
!23 = !DILocation(line: 12, column: 3, scope: !15)
!24 = !DILocation(line: 12, column: 14, scope: !15)
!25 = !DILocation(line: 12, column: 20, scope: !15)
!26 = distinct !DISubprogram(name: "println <:Bool>", linkageName: "println <:Bool>", scope: !15, file: !2, line: 13, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!27 = !DILocation(line: 13, scope: !26)
!28 = distinct !DISubprogram(name: "id <:Integer>", linkageName: "id <:Integer>", scope: !15, file: !2, line: 8, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!29 = !DILocation(line: 8, column: 19, scope: !28)
!30 = distinct !DISubprogram(name: "id <:String>", linkageName: "id <:String>", scope: !15, file: !2, line: 8, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!31 = !DILocation(line: 8, column: 19, scope: !30)
!32 = distinct !DISubprogram(name: "id <:Point<Integer>>", linkageName: "id <:Point<Integer>>", scope: !15, file: !2, line: 8, type: !4, spFlags: DISPFlagDefinition, unit: !1)
!33 = !DILocation(line: 8, column: 19, scope: !32)
